presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -963,7 +963,9 @@ else if (criteria instanceof JoinOn) {
                 // we do it further down when after we determine which subexpressions apply to left vs right tuple)
                 ExpressionAnalyzer analyzer = ExpressionAnalyzer.create(analysis, session, metadata, sqlParser, accessControl);
 
-                Type clauseType = analyzer.analyze(expression, output);
+                // need to register coercions in case when join criteria requires coercion (e.g. join on char(1) = char(2))
+                ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, output);
```

Comment:
I don't think this change is needed. 
`Type clauseType = analyzer.analyze(expression, output);`
already performs analysis (and adds it to analysis) and canonicalized expression is analyzed in:
https://github.com/prestodb/presto/pull/7529/commits/def126c9b42ff1416ecf2684fbeb03514a753688#diff-1a785e48f45c659076f1e4ec089517deR980.

The cause of the original issue was `expressionOptimizer` not using analysis, but the analysis should be there.